src/Autofac.Diagnostics.DotGraph/DotGraphBuilder.cs

```diff
@@ -0,0 +1,256 @@
+        public void OnRequestSuccess(object? instance)
+        {
+            if (CurrentRequest.Count == 0)
+            {
+                // OnRequestSuccess happened without a corresponding OnRequestStart.
+                return;
+            }
+
+            var request = Requests[CurrentRequest.Pop()];
+            request.Success = true;
+            request.Instance = instance;
+        }
+
+        /// <inheritdoc/>
+        public override string ToString()
+        {
+            var builder = new StringBuilder();
+            builder.AppendLine("digraph G {");
+            Operation.ToString(builder);
+            foreach (var request in Requests)
+            {
+                request.ToString(builder, Requests);
+            }
+
+            builder.AppendLine("}");
+            return builder.ToString();
+        }
+
+        private void NormalizeGraph()
+        {
+            // Remove any duplicates of the root node. We need to make sure that node in particular stays.
+            RemoveDuplicates(OriginatingRequest);
+
+            // Other than the originating request, find the rest of the distinct values
+            // so we can de-dupe.
+            var unique = Requests
+                .Where(r => r.Success && r.Instance is object && r.Id != OriginatingRequest)
+                .Distinct(InstanceEqualityComparer.Default)
+                .Select(r => r.Id)
+                .ToArray();
+
+            foreach (var id in unique)
+            {
+                RemoveDuplicates(id);
+            }
+        }
+
+        private void RemoveDuplicates(Guid sourceId)
+        {
+            if (!Requests.Contains(sourceId))
+            {
+                // Should always find this value, but in testing
+                // sometimes the mock value uses an empty GUID or
+                // something that might not be here. Also, it appears
+                // TryGetValue on KeyedCollection<K,V> wasn't added
+                // until netstandard2.1, so it causes compiler problems
+                // to multitarget netstandard2.0.
+                return;
+            }
+
+            var source = Requests[sourceId];
+            if (!source.Success || source.Instance is null)
+            {
+                // We can only de-duplicate successful operations because
+                // failed operations don't have instances to compare.
+                return;
+            }
+
+            var duplicates = Requests.Where(dup =>
+
+                // Successful requests where IDs are different
+                dup.Id != sourceId && dup.Success &&
+
+                // Instance is exactly the same
+                dup.Instance is object && object.ReferenceEquals(dup.Instance, source.Instance) &&
+
+                // Decorator target must also be the same (otherwise we lose the instance/decorator relationship)
+                dup.DecoratorTarget == source.DecoratorTarget).ToArray();
+            if (duplicates.Length == 0)
```

Comment:
Suggest putting a line break here to make it slightly easier on the eyes separating that block statement.

Reply:
✅